### PR TITLE
[WIP][ENG-4187] Test file detail view section - Updated

### PR DESCRIPTION
**Pull request by chth0n1x**
-   Ticket: https://openscience.atlassian.net/browse/ENG-4187
-   Feature flag: feature/file-metadata-mirage-tests

## Purpose

The purpose of these changes is to add tests for the ember-osf-web portion of the File Metadata project.

## Summary of Changes

Tests were added to the registration-file-detail-test.ts file to verify the functionality of metadata label and div elements, the edit metadata form and its save and cancel buttons, and the mobile default view for the application.

## Screenshot(s)

image

## Side Effects

There should be no unintended side effects.

## QA Notes

-Are all added components and functionalities properly tested?
-Are there any suggested tests not already covered?
-Does the default view change for tablet view? From tablet to mobile, mobile to desktop and vice versa?
-Are all elements properly rendered on the screen? Does the edit metadata button remain hidden when permissions change?

### CODE REVIEW

File: tests/acceptance/guid-file/registration-file-detail-test.ts [hunk 1]
```diff
@@ -98,7 +112,167 @@ module('Acceptance | guid file | registration files', hooks => {
...
+            // Screenshot before changes
+            await percySnapshot(assert);
+            // Update Title
+            const editTitleTextArea = document.querySelectorAll('textarea')[0];
+            editTitleTextArea.textContent = 'A test title.';
+            assert.dom('[data-test-title-field]')
+                .containsText('A test title.', 'Metadata title field updates.');
+            // Update Description
+            const editDescriptionTextArea = document.querySelectorAll('textarea')[1];
+            editDescriptionTextArea.textContent = 'A test description.';
```
Comment (brianjgeiger) [inline tests/acceptance/guid-file/registration-file-detail-test.ts:241]: Perhaps `await fillIn('[data-test-description-field]', 'A test description.');` would work here?

File: tests/acceptance/guid-file/registration-file-detail-test.ts [hunk 2]
```diff
@@ -98,7 +112,167 @@ module('Acceptance | guid file | registration files', hooks => {
...
+            assert.dom('[data-test-title-field]')
+                .containsText('A test description.', 'Metadata description field updates.');
```
Comment (brianjgeiger) [inline tests/acceptance/guid-file/registration-file-detail-test.ts:243]: Shouldn't this be 'A test title.' here, since you're verifying the title field? Why isn't this test failing?

File: tests/acceptance/guid-file/registration-file-detail-test.ts [hunk 3]
```diff
@@ -98,7 +112,167 @@ module('Acceptance | guid file | registration files', hooks => {
...
+            // Update Resource Language
+            await selectChoose('[data-test-select-resource-language]', 'InteractiveResource');
+            // Screenshot after changes
+            await percySnapshot(assert);
+            // Save changes
+            await click('[data-test-save-metadata-button]');
+        }
+        // Verify form closes
+        assert.dom('[data-test-edit-metadata-form]').doesNotExist();
+        assert.dom('[data-test-metadata-tab]').exists();
```
Comment (brianjgeiger) [inline tests/acceptance/guid-file/registration-file-detail-test.ts:255]: You should verify all of the fields properly changed after the save, not just that the edit state changed. Similarly, you should verify that cancelling the edit changes the save state without changing them values of the fields.

File: tests/acceptance/guid-file/registration-file-detail-test.ts [hunk 4]
```diff
@@ -58,16 +62,26 @@ module('Acceptance | guid file | registration files', hooks => {
...
+        assert.dom('[data-test-file-renderer]').exists();
+        assert.dom('[data-test-metadata-tab]').doesNotExist();
+        assert.dom('[data-test-revisions-tab]').doesNotExist();
+        assert.dom('[data-test-tags-tab]').doesNotExist();
+
+        // click tags and verify appropriate panels shown
         await click('[data-test-versions-button]');
         assert.dom('[data-test-revisions-tab]').exists('Revisions shown');
-        assert.dom('[data-test-file-renderer]').doesNotExist('File renderer is hidden');
+        assert.dom('[data-test-file-renderer]').isNotVisible('File renderer is hidden');
```
Comment (brianjgeiger) [inline tests/acceptance/guid-file/registration-file-detail-test.ts:73]: I thought we decided to use `.doesNotExist` rather than `.isNotVisible` unless `.doesNotExist` just doesn't work. Here and below.

File: tests/acceptance/guid-file/registration-file-detail-test.ts [hunk 5]
```diff
@@ -98,7 +112,167 @@ module('Acceptance | guid file | registration files', hooks => {
...
+                .containsText('A test title.', 'Metadata title field updates.');
+            // Update Description
+            const editDescriptionTextArea = document.querySelectorAll('textarea')[1];
+            editDescriptionTextArea.textContent = 'A test description.';
+            assert.dom('[data-test-title-field]')
+                .containsText('A test description.', 'Metadata description field updates.');
+            // Update Resource Type
+            await selectChoose('[data-test-select-resource-type]', 'InteractiveResource');
+            // Update Resource Language
+            await selectChoose('[data-test-select-resource-language]', 'InteractiveResource');
```
Comment (brianjgeiger) [inline tests/acceptance/guid-file/registration-file-detail-test.ts:247]: This is not a valid choice for languages.

File: tests/acceptance/guid-file/registration-file-detail-test.ts [hunk 6]
```diff
@@ -98,7 +112,167 @@ module('Acceptance | guid file | registration files', hooks => {
...
+    test('view metadata', async function(this: ThisTestContext, assert) {
+        const nodeNoun = ['Registration', 'Project', 'Component'];
+        const metadataRecord = await this.owner.lookup('service:store')
+            .findRecord('custom-item-metadata-record', this.registration.id);
+
+        await visit(`/--file/${this.file.id}`);
+        assert.equal(currentURL(), `/--file/${this.file.guid}`);
+
+        // Verify correct default tab for dekstop
+        assert.dom('[data-test-metadata-tab]').exists();
```
Comment (brianjgeiger) [inline tests/acceptance/guid-file/registration-file-detail-test.ts:125]: This test selector does not exist, so you have a few failing tests.

File: tests/acceptance/guid-file/registration-file-detail-test.ts [hunk 7]
```diff
@@ -98,7 +112,167 @@ module('Acceptance | guid file | registration files', hooks => {
...
+        assert.equal(currentURL(), `/--file/${this.file.guid}`);
+
+        // Verify download
+        if (!this.registration.isAnonymous) {
+            assert.dom('[data-test-download-button]').exists();
+        }
 
-    // test for switching views
+        // Verify edit
+        if (this.file.target.currentUserPermissions.includes(Permission.Write)) {
```
Comment (brianjgeiger) [inline tests/acceptance/guid-file/registration-file-detail-test.ts:220]: This if statement, also, is not appropriate. You should know what your current user's permissions are and set them to be the appropriate permissions so that this these assertions will run.

File: tests/acceptance/guid-file/registration-file-detail-test.ts [hunk 8]
```diff
@@ -98,7 +112,167 @@ module('Acceptance | guid file | registration files', hooks => {
...
+        assert.dom('[data-test-target-description]').exists();
+    });
+
+    // Save and cancel buttons with write permissions
+    test('save and cancel', async function(this: ThisTestContext, assert) {
+        await visit(`/--file/${this.file.id}`);
+        assert.equal(currentURL(), `/--file/${this.file.guid}`);
+
+        // Verify download
+        if (!this.registration.isAnonymous) {
```
Comment (brianjgeiger) [inline tests/acceptance/guid-file/registration-file-detail-test.ts:215]: This if statement also should not exist.

**Review by brianjgeiger — CHANGES_REQUESTED** [review]
Okay, it looks like you're going to need some more test scenarios to test more cases. Also, there's some inappropriate conditional logic in there, and one of the tests isn't testing the way I think it should.

File: tests/acceptance/guid-file/registration-file-detail-test.ts (hunk 6 above)
Comment (chth0n1x) [reply tests/acceptance/guid-file/registration-file-detail-test.ts:125]: Ok, this has been added.

File: tests/acceptance/guid-file/registration-file-detail-test.ts (hunk 7 above)
Comment (chth0n1x) [reply tests/acceptance/guid-file/registration-file-detail-test.ts:220]: Ok, this is removed.

File: tests/acceptance/guid-file/registration-file-detail-test.ts (hunk 8 above)
Comment (chth0n1x) [reply tests/acceptance/guid-file/registration-file-detail-test.ts:215]: Ok, this is removed.

File: tests/acceptance/guid-file/registration-file-detail-test.ts (hunk 5 above)
Comment (chth0n1x) [reply tests/acceptance/guid-file/registration-file-detail-test.ts:247]: Per our Slack conversation, selectChoose will remain.

File: tests/acceptance/guid-file/registration-file-detail-test.ts (hunk 4 above)
Comment (chth0n1x) [reply tests/acceptance/guid-file/registration-file-detail-test.ts:73]: This is now updated for this test as well.

File: tests/acceptance/guid-file/registration-file-detail-test.ts (hunk 5 above)
Comment (brianjgeiger) [reply tests/acceptance/guid-file/registration-file-detail-test.ts:247]: But, also per our slack conversation 'InteractiveResource' should be replaced with an actual language.

File: tests/acceptance/guid-file/registration-file-detail-test.ts (hunk 7 above)
Comment (chth0n1x) [reply tests/acceptance/guid-file/registration-file-detail-test.ts]: This is now a permission on the registration instead of a conditional in the test.

File: tests/acceptance/guid-file/registration-file-detail-test.ts (hunk 2 above)
Comment (chth0n1x) [reply tests/acceptance/guid-file/registration-file-detail-test.ts]: This is now working by updating the data test selector.

File: tests/acceptance/guid-file/registration-file-detail-test.ts (hunk 1 above)
Comment (chth0n1x) [reply tests/acceptance/guid-file/registration-file-detail-test.ts]: This is now working.

File: tests/acceptance/guid-file/registration-file-detail-test.ts [hunk 9]
```diff
@@ -98,7 +112,167 @@ module('Acceptance | guid file | registration files', hooks => {
...
+            assert.dom('[data-test-target-contributors-label]').hasText('Contributors',
+                'File metadata contributor text is properly set.');
+            assert.dom('[data-test-target-contributors]').exists();
+            // Target Institutions
+            assert.dom('[data-test-target-institutions-label]').hasText('Affiliated institutions',
+                'File affiliated institutions text is properly set.');
+            assert.dom('[data-test-target-institutions]').exists();
+        }
+        // Funder metadata
+        if (!this.registration.isAnonymous && metadataRecord.funder) {
```
[the remark this answers is not in the document]
Comment (chth0n1x) [reply tests/acceptance/guid-file/registration-file-detail-test.ts:174]: This is now removed.

File: tests/acceptance/guid-file/registration-file-detail-test.ts (hunk 3 above)
Comment (brianjgeiger) [reply tests/acceptance/guid-file/registration-file-detail-test.ts:277]: Technically, this is an expansion from the original ticket's scope. When the original ticket was made, we only had view functionality and not edit functionality, so we were just testing what we had at the time. I had forgotten about that over the course of time. I suppose we can pull this off onto a separate ticket.

File: tests/acceptance/guid-file/registration-file-detail-test.ts [hunk 10]
```diff
@@ -98,7 +112,167 @@ module('Acceptance | guid file | registration files', hooks => {
...
+        // Resource language
+        assert.dom('[data-test-file-language-label]').hasText('Resource language',
+            'File metadata resource language text is properly set.');
+        assert.dom('[data-test-file-language]').exists();
+
+        // Node metadata
+        const nodeNounElement = document.querySelectorAll('[data-test-metadata-node]')[0].textContent;
+        const nounFound = nodeNoun.some(r => nodeNounElement?.includes(r));
+        assert.equal(nounFound, true, 'Node noun properly set.');
+        if (!this.registration.isAnonymous && (this.registration.contributors.length !== 0)) {
```
[the remark this answers is not in the document]
Comment (brianjgeiger) [reply tests/acceptance/guid-file/registration-file-detail-test.ts:163]: Similarly, we'll hold off on testing anonymous items, since those weren't available when the original ticket was written.

File: tests/acceptance/guid-file/registration-file-detail-test.ts [hunk 11]
```diff
@@ -98,7 +118,196 @@ module('Acceptance | guid file | registration files', hooks => {
...
+        await selectChoose('[data-test-select-resource-language]', 'Polish');
+        await selectChoose('[data-test-select-resource-language]', 'Portuguese');
+        await selectChoose('[data-test-select-resource-language]', 'Russian');
+        await selectChoose('[data-test-select-resource-language]', 'South American Indian languages');
+        await selectChoose('[data-test-select-resource-language]', 'Spanish');
+        await selectChoose('[data-test-select-resource-language]', 'Swedish');
+        await selectChoose('[data-test-select-resource-language]', 'Tagalog');
+        await selectChoose('[data-test-select-resource-language]', 'Thai');
+        await selectChoose('[data-test-select-resource-language]', 'Vietnamese');
+        await selectChoose('[data-test-select-resource-language]', 'Zuni');
```
Comment (brianjgeiger) [inline tests/acceptance/guid-file/registration-file-detail-test.ts:271]: Out of curiosity, why did you select 23 different languages here?

File: tests/acceptance/guid-file/registration-file-detail-test.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (chth0n1x) [reply tests/acceptance/guid-file/registration-file-detail-test.ts:164]: The anonymous test is passing and is now included.

File: tests/acceptance/guid-file/registration-file-detail-test.ts [hunk 12]
```diff
@@ -98,7 +118,196 @@ module('Acceptance | guid file | registration files', hooks => {
...
+    test('anonymous registration', async function(this: ThisTestContext, assert) {
+        const node = server.create('node', {
+            isAnonymous: true,
+        });
+
+        await visit(`/--node/${node.id}`);
+        assert.equal(currentURL(), `/--node/${node.id}`);
+
+        assert.dom('[data-test-metadata-node]').doesNotExist();
+    });
```
Comment (brianjgeiger) [inline tests/acceptance/guid-file/registration-file-detail-test.ts:291]: This test doesn't work. The method to create an anonymous node is slightly different from this. 

The reason the assertion passes is because you're going to a page that doesn't have the metadata on it and checking for the non-existence a data-test selector that is unique to the metadata section. Best to just remove this for now and save it for a future ticket.

File: tests/acceptance/guid-file/registration-file-detail-test.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (chth0n1x) [reply tests/acceptance/guid-file/registration-file-detail-test.ts:160]: The test for registration node noun is now also added. We may want to include the tests for components and projects in another ticket since from our conversation this is slightly out of scope. In the meantime, I can include my check against the node noun array for Registration, Project, and Component back if you think it helps. Let me know which you would prefer.

File: tests/acceptance/guid-file/registration-file-detail-test.ts [hunk 13]
```diff
@@ -98,7 +112,167 @@ module('Acceptance | guid file | registration files', hooks => {
...
+        assert.dom('[data-test-file-resource-type]').exists();
+        // Resource language
+        assert.dom('[data-test-file-language-label]').hasText('Resource language',
+            'File metadata resource language text is properly set.');
+        assert.dom('[data-test-file-language]').exists();
+
+        // Node metadata
+        const nodeNounElement = document.querySelectorAll('[data-test-metadata-node]')[0].textContent;
+        const nounFound = nodeNoun.some(r => nodeNounElement?.includes(r));
+        assert.equal(nounFound, true, 'Node noun properly set.');
```
[the remark this answers is not in the document]
Comment (chth0n1x) [reply tests/acceptance/guid-file/registration-file-detail-test.ts]: A separate test for a registration node noun has now replaced the previous one. We may want to include the tests for components and projects in another ticket since from our conversation this is slightly out of scope. In the meantime, I can include my check against the node noun array for Registration, Project, and Component back if you think it helps. Let me know which you would prefer.

[the remark this answers is not in the document]
Comment (brianjgeiger) [reply tests/acceptance/guid-file/registration-file-detail-test.ts]: No, that one is in scope. On the ticket, it was the second bullet point, "Ensure that the label for the node section correctly populates project, component, or registration. "

File: tests/acceptance/guid-file/registration-file-detail-test.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (chth0n1x) [reply tests/acceptance/guid-file/registration-file-detail-test.ts]: The test is also now passing with contributors and affiliated institutions.

File: tests/acceptance/guid-file/registration-file-detail-test.ts (hunk 11 above)
Comment (chth0n1x) [reply tests/acceptance/guid-file/registration-file-detail-test.ts:276]: It was the most complete without testing edge cases for content uploaded. If I have missed any you would like to see included, please let me know.

File: tests/acceptance/guid-file/registration-file-detail-test.ts (hunk 12 above)
Comment (chth0n1x) [reply tests/acceptance/guid-file/registration-file-detail-test.ts]: Ok, removing this to be included in a future ticket.

File: tests/acceptance/guid-file/registration-file-detail-test.ts (hunk 11 above)
Comment (brianjgeiger) [reply tests/acceptance/guid-file/registration-file-detail-test.ts:276]: I think just picking one language will do fine for this test. If there were edges to test, then it would make sense to test those edges, but in this case, just ensuring that the choose succeeds should be enough.

File: tests/acceptance/guid-file/registration-file-detail-test.ts [hunk 14]
```diff
@@ -98,7 +122,185 @@ module('Acceptance | guid file | registration files', hooks => {
...
+        assert.dom('[data-test-target-description]').exists();
+    });
+
+    // Node type
+    test('Node type', async function(this: ThisTestContext, assert) {
+        await visit(`/--file/${this.file.id}`);
+        assert.equal(currentURL(), `/--file/${this.file.guid}`);
+
+        assert.dom('[data-test-metadata-node]').hasText('Registration Metadata',
+            'Node noun for registration properly set.');
```
Comment (brianjgeiger) [inline tests/acceptance/guid-file/registration-file-detail-test.ts:216]: As mentioned [here](https://github.com/CenterForOpenScience/ember-osf-web/pull/1793#discussion_r1102034616), the other two versions of this test are within scope, as it was requested in the original ticket.

**Review by brianjgeiger — CHANGES_REQUESTED** [review]
Nearly there. Just three things left:
1. Remove another if statement
2. Remove most of the language selections
3. Add in the tests for node word in the case of project and component, as they were requested in the original ticket.

File: tests/acceptance/guid-file/registration-file-detail-test.ts (hunk 11 above)
Comment (chth0n1x) [reply tests/acceptance/guid-file/registration-file-detail-test.ts:276]: Ok, I will test in English then. Thank you.

File: tests/acceptance/guid-file/registration-file-detail-test.ts [hunk 15]
```diff
@@ -98,7 +122,185 @@ module('Acceptance | guid file | registration files', hooks => {
...
+        // Contributor metadata
+        assert.dom('[data-test-target-contributors-label]').hasText('Contributors',
+            'File metadata contributor text is properly set.');
+        assert.dom('[data-test-target-contributors]').exists();
+        // Target institutions
+        assert.dom('[data-test-target-institutions-label]').hasText('Affiliated institutions',
+            'File affiliated institutions text is properly set.');
+        assert.dom('[data-test-target-institutions]').exists();
+        // Funder metadata
+        if (metadataRecord.funder) {
```
[the remark this answers is not in the document]
Comment (chth0n1x) [reply tests/acceptance/guid-file/registration-file-detail-test.ts:171]: Ok, removing the conditional statement.

File: tests/acceptance/guid-file/registration-file-detail-test.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (chth0n1x) [reply tests/acceptance/guid-file/registration-file-detail-test.ts]: Yes, it was just resolved differently. I have it checking values for each Project, Component, or Registration but understand you would like to see the test constructed for a top-level structure that navigates to each and verifies if the label states it is a Project's, Component's, or Registration's metadata. This can also be added.

**Comment by adlius** [comment]
Is this a duplicate of #1803?

**Comment by brianjgeiger** [comment]
Replaced by https://github.com/CenterForOpenScience/ember-osf-web/pull/1802
